service/src/main/java/bio/terra/externalcreds/services/OAuth2Service.java

```diff
@@ -0,0 +1,138 @@
+package bio.terra.externalcreds.services;
+
+import java.time.Instant;
+import java.util.Map;
+import java.util.Set;
+import org.springframework.security.oauth2.client.endpoint.DefaultAuthorizationCodeTokenResponseClient;
+import org.springframework.security.oauth2.client.endpoint.DefaultRefreshTokenTokenResponseClient;
+import org.springframework.security.oauth2.client.endpoint.OAuth2AuthorizationCodeGrantRequest;
+import org.springframework.security.oauth2.client.endpoint.OAuth2RefreshTokenGrantRequest;
+import org.springframework.security.oauth2.client.registration.ClientRegistration;
+import org.springframework.security.oauth2.client.userinfo.DefaultOAuth2UserService;
+import org.springframework.security.oauth2.client.userinfo.OAuth2UserRequest;
+import org.springframework.security.oauth2.core.OAuth2AccessToken;
+import org.springframework.security.oauth2.core.OAuth2RefreshToken;
+import org.springframework.security.oauth2.core.endpoint.OAuth2AccessTokenResponse;
+import org.springframework.security.oauth2.core.endpoint.OAuth2AuthorizationExchange;
+import org.springframework.security.oauth2.core.endpoint.OAuth2AuthorizationRequest;
+import org.springframework.security.oauth2.core.endpoint.OAuth2AuthorizationResponse;
+import org.springframework.security.oauth2.core.user.OAuth2User;
+import org.springframework.stereotype.Service;
+
+/**
+ * Service that encapsulates all OAuth2 features need by ECM. General flow: 1)
+ * getAuthorizationRequestUri - visit to authenticate with provider 2) authorizationCodeExchange -
+ * using code resulting from authentication, get access and refresh tokens 3) getUserInfo - using
+ * access token call the user info endpoint - called periodically to get updated user info 4)
+ * authorizeWithRefreshToken - get a new access token using a refresh token when required
```

Comment:
To make intellij format it nicely, here's the version with the proper list, as returned after passing it through the formatter, which insists that list items don't need close tags:
```suggestion
 * Service that encapsulates all OAuth2 features need by ECM. General flow:
 *
 * <ol>
 *   <li>getAuthorizationRequestUri - visit to authenticate with provider
 *   <li>authorizationCodeExchange - using code resulting from authentication, get access and
 *       refresh tokens
 *   <li>getUserInfo - using access token call the user info endpoint - called periodically to get
 *       updated user info
 *   <li>authorizeWithRefreshToken - get a new access token using a refresh token when required
 * </ol>
```